Security

dmarmul opened the pull request:
Used spring-boot-starter-security.
Added SecurityConfig class.
Implemented UserDetails interface in the User class. Added UserDetailsService implementation.
Made /auth and swagger public available.
Added Role entity and RoleRepository.
Added Set<Role> roles to User. Made role 'user' default in UserServiceImpl Inserted roles and user-roles tables into DB and made 1 admin user. Added @PreAuthorize annotation for adding a book.
Added hashing password when user register
disable liquibase in test class because of 'liquibase referential integrity violation' when use 'mvn clean package'

dmarmul committed:
Used spring-boot-starter-security.
Added SecurityConfig class.
Implemented UserDetails interface in the User class.
Added UserDetailsService implementation.
Made /auth and swagger public available.
Added Role entity and RoleRepository.
Added Set<Role> roles to User. Made role 'user' default in UserServiceImpl
Inserted roles and user-roles tables into DB and made 1 admin user.
Added @PreAuthorize annotation for adding a book.
Added hashing password when user register
disable liquibase in test class because of 'liquibase referential integrity violation' when use 'mvn clean package'

Elena-Bruyako commented on src/main/java/org/example/bookshop/config/SecurityConfig.java:
```diff
@@ -0,0 +1,41 @@
+public class SecurityConfig {
+    private final UserDetailsService userDetailsService;
+
+    public SecurityConfig(UserDetailsService userDetailsService) {
```
Use Lombok annotation to generate constructor

Elena-Bruyako commented on src/main/java/org/example/bookshop/controller/BookController.java:
```diff
@@ -39,6 +43,7 @@ public BookDto getBookById(@PathVariable Long id) {
     }
 
     @PostMapping
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
add PreAuthorize for other required endpoints

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/Role.java:
```diff
@@ -0,0 +1,33 @@
+    private Long id;
+    @Column(nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private Name name;
```
```suggestion
    private Role name;
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/Role.java:
```diff
@@ -0,0 +1,33 @@
+    private Name name;
+
+    @Getter
+    public enum Name {
```
```suggestion
    public enum Role {
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/Role.java:
```diff
@@ -0,0 +1,33 @@
+@Setter
+@ToString
+@Table(name = "roles")
+public class Role {
```
```suggestion
public class Role implements GrantedAuthority {
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/Role.java:
```diff
@@ -0,0 +1,33 @@
+@Entity
+@Getter
+@Setter
+@ToString
```
```suggestion
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/User.java:
```diff
@@ -34,4 +43,46 @@ public class User {
     private String shippingAddress;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.LAZY)
```
```suggestion
    @ManyToMany
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/User.java:
```diff
@@ -34,4 +43,46 @@ public class User {
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
```suggestion
    private Set<Role> roles = new HashSet<>();
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/User.java:
```diff
@@ -34,4 +43,46 @@ public class User {
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return List.of();
```
```suggestion
        return roles;
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/repository/UserRepository.java:
```diff
@@ -1,8 +1,11 @@
 public interface UserRepository extends JpaRepository<User, Long> {
     boolean existsByEmail(String email);
+
+    Optional<User> findByEmail(String email);
```
```suggestion
    @EntityGraph(attributePaths = "roles")
    Optional<User> findByEmail(String email);
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/security/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new UsernameNotFoundException("Can't find user " + email));
```
```suggestion
            {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new EntityNotFoundException("Can't find user " + email));
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/BookServiceImpl.java:
```diff
@@ -31,6 +31,11 @@ public BookDto findById(Long id) {
         return bookMapper.toDto(book);
     }
 
+    @Override
+    public List<BookDto> findAll(String email, Pageable pageable) {
+        return List.of();
+    }
+
```
```suggestion
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/BookService.java:
```diff
@@ -12,6 +12,8 @@ public interface BookService {
 
     List<BookDto> findAll(Pageable pageable);
 
+    List<BookDto> findAll(String email, Pageable pageable);
+
```
```suggestion
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java:
```diff
@@ -23,6 +29,8 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) {
         }
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
+        user.setRoles((Set.of(roleRepository.findByName(Role.Name.USER))));
```
```suggestion
        user.setRoles(Set.of(roleRepository.findByName(Role.Name.USER)));
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java:
```diff
@@ -1,20 +1,26 @@
     private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
 
     @Override
```
```suggestion
    @Override
    @Transactional
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/controller/BookController.java:
```diff
@@ -28,8 +31,9 @@ public class BookController {
             description = "Get a list of all available books")
-    public List<BookDto> getAll(Pageable pageable) {
-        return bookService.findAll(pageable);
+    public List<BookDto> getAll(Authentication authentication, Pageable pageable) {
+        User user = (User) authentication.getPrincipal();
```
```suggestion
    public List<BookDto> getAll(Pageable pageable) {
```

Elena-Bruyako reviewed (changes requested):
see comments

dmarmul committed:
added PreAuthorize to all book endpoints.
Renamed enum class Name to RoleType, can't rename as Role, because already have this class. Implemented GrantedAuthority in Role class.
Replaced FetchType.LAZY in @ManyToMany annotation. Added @EntityGraph with roles to findByEmail method.
Removed don't used findAll method.
Added @Transactional to register method.
And fix few more minor errors.

liliia-ponomarenko commented on src/main/java/org/example/bookshop/security/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,20 @@
+    @Override
+    public UserDetails loadUserByUsername(String email) {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find user " + email));
```
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Can't find user by email:" + email));
```

liliia-ponomarenko commented on src/main/java/org/example/bookshop/model/Role.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public String getAuthority() {
+        return role.name();
```
Does it work correctly without `"ROLE_" + role.name()` ?

dmarmul replied:
should work correctly, I have 0 login method, can't check it yet. when use basic autentificstion in postman - have 403 response. try to use return (role.name();) or ("ROLE_USER") or ("USER")

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/05-create-user-roles-table.yaml:
```diff
@@ -0,0 +1,35 @@
+        - addPrimaryKey:
+            tableName: user_roles
+            columnNames: user_id, role_id
+        - addForeignKeyConstraint:
+            baseTableName: user_roles
+            baseColumnNames: user_id
+            constraintName: fk_user_roles_user
+            referencedTableName: users
+            referencedColumnNames: id
+            onDelete: CASCADE
+        - addForeignKeyConstraint:
+            baseTableName: user_roles
+            baseColumnNames: role_id
+            constraintName: fk_user_roles_role
+            referencedTableName: roles
+            referencedColumnNames: id
+            onDelete: CASCADE
```
I can't find the same cascade in the entity class

dmarmul replied:
I read that if need to remove from db, only liqubase is enough. and it's working, when I remove from db by query console. For example, when I remove user from db, from user_roles also remove this user and it's role

Elena-Bruyako replied:
check same logic not from console

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/07-insert-admin-to-user-roles-table.yaml:
```diff
@@ -0,0 +1,10 @@
+                  tableName: user_roles
+                  columns:
+                    - column: { name: user_id, value: 1 }
+                    - column: { name: role_id, value: 2 }
```
add one empty line at the end of the file (check other files)

liliia-ponomarenko commented on src/test/resources/application.properties:
```diff
@@ -3,3 +3,4 @@ spring.datasource.driverClassName=org.h2.Driver
 spring.datasource.username=sa
 spring.datasource.password=password
 spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
+spring.liquibase.enabled=false
```
why do you disable liquibase?

dmarmul replied:
disable liquibase in test class because of 'liquibase referential integrity violation' when use 'mvn clean package'.  Нарушение ссылочной целостности: "FK_USER_ROLES_ROLE: PUBLIC.USER_ROLES FOREIGN KEY(ROLE_ID) REFERENCES PUBLIC.ROLES(ID) (CAST(2 AS BIGINT))"
Referential integrity constraint violation: "FK_USER_ROLES_ROLE: PUBLIC.USER_ROLES FOREIGN KEY(ROLE_ID) REFERENCES PUBLIC.ROLES(ID) (CAST(2 AS BIGINT))"; SQL statement:
INSERT INTO PUBLIC.user_roles (user_id, role_id) VALUES ('1', '2') [23506-224] [Failed SQL: (23506) INSERT INTO PUBLIC.user_roles (user_id, role_id) VALUES ('1', '2')]

Elena-Bruyako replied:
try to resolve this error withour disabling liquibase in test

liliia-ponomarenko reviewed (changes requested):
Good job! A few comments. Please don't resolve comments. You can use emoji for marking fixed comments)

dmarmul committed:
Sorry, wrote in the comments for every comment I didn't fix, I think the part will be solved when the login method is added

dmarmul committed:
Added .httpBasic(Customizer.withDefaults()) and authentication started to work correctly. Used PreAuthorize role without ROLE_ prefix. Added ROLE_ prefix to getAuthority method return

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/Role.java:
```diff
@@ -0,0 +1,39 @@
+        return AUTHORITY_PREFIX + role.name();
+    }
+
+    @Getter
```
```suggestion
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/model/User.java:
```diff
@@ -34,4 +42,46 @@ public class User {
     private boolean isDeleted = false;
+    @ManyToMany
+    @JoinTable(
+            name = "user_roles",
```
```suggestion
            name = "users_roles",
```

Elena-Bruyako commented on src/main/resources/db/changelog/changes/05-create-user-roles-table.yaml:
```diff
@@ -0,0 +1,35 @@
+      author: mateacademystudent
+      changes:
+        - createTable:
+            tableName: user_roles
```
```suggestion
            tableName: users_roles
```

Elena-Bruyako reviewed (changes requested):
see comments

dmarmul committed:
Removed Getter annotation from Enum class.
Rename user_roles table.
"onDelete: CASCADE" is delete user_id and role_id from table users_roles when use postman also(even when in users table stay deleted user with changed value in 'is_deleted' column.
Changed liquibase files and deleted ignore liquibase properties from application.properties in test folder

ahoienko reviewed (approved):
Good job!
